### ISSUE TITLE
Accessibility testing

### ISSUE THREAD

**Issue by rosejbon** [neontribe/ca-hospice-hope#57]
## Description
Review the accessibility of the app

## Basic Accessibility Checks
- [x] **Non-text Content** - There are text alternatives for all non-text content https://www.wuhcag.com/non-text-content/
- [ ] **Logical Structure** - All HTML is correct and valid (with no errors) https://www.wuhcag.com/info-and-relationships/
- [ ] **Meaningful Sequence** - Content is presented in a meaningful order https://www.wuhcag.com/meaningful-sequence/
- [x] **Use of Colour** - Instructions don't rely on colour alone https://www.wuhcag.com/use-of-colour/
- [x] **Keyboard** - When mouse is unplugged, I can use all features of the site https://www.wuhcag.com/keyboard/
- [x] **No Keyboard Trap** - All navigation is controllable by the 'tab' key https://www.wuhcag.com/no-keyboard-trap/
- [ ] **Page Titles** - All titles are clear and explain the page well https://www.wuhcag.com/page-titled/
- [ ] **Focus Order** - Tabbing through the page flows nicely and is in a logical order https://www.wuhcag.com/focus-order/  
- [ ] **Link text** - The purpose of every link is clear from its text/context https://www.wuhcag.com/link-purpose-in-context/
- [x] **Language of a page** - Each page of the site has a language assigned to it https://www.wuhcag.com/language-of-page/
- [x] **On Focus** - When there is focus on buttons and links they don't autosubmit https://www.wuhcag.com/on-focus/
- [x] **On Input** - Focus shouldn't unnecessarily shift between inputs https://www.wuhcag.com/on-input/
- [x] **Labels** - All input fields are labelled clearly and helpfully https://www.wuhcag.com/labels-or-instructions/

## Intermediate Accessibility Checks 

- [ ] **Resize Text** - Text can be resized to 200% without loss of content/function https://www.wuhcag.com/resize-text/
- [ ] **Focus Visible** - Keyboard focus is visible on all all elements https://www.wuhcag.com/focus-visible/
- [ ] **Colour Contrast** - Contrast ratio of text & background at least 4.5:1 (can test with Tota11y) https://www.wuhcag.com/contrast-minimum/
- [ ] **Consistent Navigation** - Navigation is in the same location on all pages https://www.wuhcag.com/consistent-navigation/
- [ ] **Consistent Identification** - Elements with equal function are consistent https://www.wuhcag.com/consistent-identification/
- [ ] **Error Hints** - Fixes are suggested when users make errors https://www.wuhcag.com/error-suggestion/

**Comment by Hoolean** [comment]
I will be working through these points. For each, I will be testing the Hospice Hope site (but not the linked contact form) and either ticking as correct or documenting failings below. If a point is unticked and there is no reason given, I haven't gotten to it yet.

**Comment by Hoolean** [comment]
* [x] **Logical Structure**/**Meaningful Sequence**

  These findings should be considered preliminary; this will be hard to confirm as ready without screenreader testing.

  The layout of pages is conducive to a user seeking support. HTML5 use is generally semantic. A few potential exceptions are documented here:
    * [x] Use of headings for long Call To Action text can make the outline that a screenreader would use bloated. e.g.
    ![image](https://user-images.githubusercontent.com/2150762/63609898-02a8d800-c5d0-11e9-94a8-f60f57b8499a.png)
    * [x] Some sections, while made distinct with visual styling and HTML5 `<section>` tags, are left without headings. This leaves them invisible and unlabelled in the document outline. For each section, it should be considered as to whether they would be deserving of a place in the outline. e.g.
      * The information given after a story:
        ![image](https://user-images.githubusercontent.com/2150762/63610714-fd4c8d00-c5d1-11e9-91b9-f6bf62d05759.png)
      * The dates given within "When do sessions take place?"
    * [x] Some areas may benefit from being explicitly ordered or marked as unordered, to aid screenreader navigation features. e.g. the footer is only sectioned with `<div>` elements
    ![image](https://user-images.githubusercontent.com/2150762/63610219-d3469b00-c5d0-11e9-873b-223cccc410b5.png)

* [x] **Page Titles** - All titles are clear and explain the page well https://www.wuhcag.com/page-titled/

  As of right now, the page title (as seen in the window title) stays constant for every page. This is undesirable.

* [x] **Focus Order** - Tabbing through the page flows nicely and is in a logical order https://www.wuhcag.com/focus-order/

  The order of tabbing is consistently good: the header, the navbar (whether at the top or bottom) and then the content.

  Despite this, the site is not consistently amenable to tab-based navigation:
  * [ ] If the browser must scroll to reach the next keyboard-navigable element, it may be covered by the sticky navbar on small-viewport displays. e.g. The `supporters` link has been selected and automatically scrolled to but is obscured (on Firefox 68.0.2, Ubuntu 19.04).
    ![image](https://user-images.githubusercontent.com/2150762/63600196-00884e80-c5bb-11e9-83dd-df3a9cbe1b8e.png)
  * [ ] The site uses four separate links for the same story in each card, all of which must be tabbed through individually. This also presents a hindrance for users of screenreaders.
    ![image](https://user-images.githubusercontent.com/2150762/63600711-fc106580-c5bb-11e9-8d29-44ba152807b8.png)
  * [ ] There is an empty link that must be tabbed past on the homepage. As it is not described, this would be especially confusing for users of screenreaders.
    ![image](https://user-images.githubusercontent.com/2150762/63600867-4abdff80-c5bc-11e9-8e54-5d7fd716defc.png)
  * [ ] There is an empty heading on the Privacy Policy page
    ![image](https://user-images.githubusercontent.com/2150762/63611379-aba50200-c5d3-11e9-8869-97b054c0b2cc.png)


  Aside from these flaws, I have a few other open questions:
  * Should the cookie banner be navigated to by tab before anything else? The generally accessible [BBC website](https://bbc.co.uk/) does prioritise the banner, so maybe this is not too much of a crime. Assuming the choice is meant to be made once at the beginning of a session, this doesn't seem too troublesome. Still, it is undeniably more of a barrier in the linear style of browsing a screenreader is limited to.
  * [ ] Typically, when a user navigates to a new page, tab-navigation begins from the root again. As Contemplating Action is a Single Page App, our behaviour is not so obvious. From experimentation, our app appears relatively well-behaved, apparently [focusing at the start of the new content (linked and listed here as strategy two)](https://daverupert.com/2019/01/accessible-page-navigations-in-single-page-apps/). It would be reassuring to have confirmed that this behaviour is intentional and reliable.
    **EDIT: Following further testing, I have found that following a link in the footer will leave the tab navigation positioned relative to there, even though the page has changed. In addition, following a link to the same page will not cause a refresh and will leave the link focused. We should investigate this behaviour further.**

* [x] **Link text** - The purpose of every link is clear from its text/context https://www.wuhcag.com/link-purpose-in-context/

  Most links are clear and obvious (aside from the empty link that goes nowhere, as mentioned in **Focus Order**). There are still a few that are worth discussing:
  * [ ] In the cookie banner, "Accept" and "Decline" may not be obvious in their effect if read by a screenreader out of context. Links may be read out of context for speedy navigation, so this is not too uncommon a scenario.
  * [ ] "Show me the stories" and "Hope Stories" on the homepage have different descriptions but both link to the same (which is advised against in the guidelines of the website linked).
  * [ ] The copyright URL in the footer and multiple URLs in the privacy policy have no human-readable description; a screen reader would just have to read them out in full.

* [x] **Other**
  * [ ] Quotes in stories are given their quotation marks with CSS pseudo-elements. This may not work smoothly with screenreaders.
  ![image](https://user-images.githubusercontent.com/2150762/63619441-908fbd80-c5e6-11e9-8e72-9dd0bef25720.png)

**Comment by Hoolean** [comment]
I imagine Typeform will just need practical testing (with screenreaders, different zooms and tab navigation), as we cannot change anything at the code level, only more to another platform. 

Previous opinions on Typeform's accessibility [have not been very laudatory](https://a11y.reviews/#typeform) but as we are only using a small subset of features, it may be sufficient.

**Comment by rosejbon** [comment]
Thanks @Hoolean, I have carded all these issues for discussion

**Comment by rosejbon** [comment]
Closing because carded
